extensions/eclipse-che-theia-plugin-remote/src/node/plugin-remote-node-impl.ts

```diff
@@ -0,0 +1,253 @@
+            await this.sleep(50);
+            waited++;
+            plugin = pluginManagerInternal.registry.get(pluginId);
+        }
+        if (waited >= 100) {
+            throw new Error(`Unable to load the plug-in ${pluginId}. Wait for being ready but was never loaded.`);
+        }
+        await pluginManagerInternal.loadPlugin(plugin, configStorage);
+        const activatedPlugin = pluginManagerInternal.activatedPlugins.get(pluginId);
+
+        // share the JSON with others
+        const rawModel = plugin.rawModel;
+        await this.pluginRemoteBrowser.$definePluginPackage(pluginId, rawModel);
+
+        if (activatedPlugin && activatedPlugin.exports) {
+
+            // do we have prototype ?
+            const exported = activatedPlugin.exports;
+
+            const prototype = Object.getPrototypeOf(exported);
+            let proxyNames: Array<string> = [];
+            if (prototype) {
+                proxyNames = proxyNames.concat(Object.getOwnPropertyNames(prototype));
+            }
+
+            // ok now need to send that back only if there are some exports
+            if (proxyNames.length > 0) {
+                await this.pluginRemoteBrowser.$definePluginExports(pluginId, proxyNames);
+            }
+        }
+    }
+
+    async $activatePlugin(pluginId: string): Promise<void> {
+        // tslint:disable-next-line: no-any
+        const pluginManagerInternal = (this.pluginManager as any);
+        await pluginManagerInternal.activatePlugin(pluginId);
+    }
+
+    async $definePluginPackage(pluginId: string, pluginPackage: PluginPackage): Promise<void> {
+        // tslint:disable-next-line: no-any
+        const plugin = (this.pluginManager as any).registry.get(pluginId);
+        if (plugin) {
+            plugin.rawModel = pluginPackage;
+        }
+    }
+
+    // tslint:disable-next-line: no-any
+    deasyncPromise(promise: Promise<any>) {
+        // tslint:disable-next-line: no-any
+        let result: any;
+        // tslint:disable-next-line: no-any
+        let error: any;
+        let done = false;
+        promise.then(res => {
+            result = res;
+        }, err => {
+            error = err;
+        }).then(() => {
+            done = true;
+        });
+        deasync.loopWhile(() => !done);
+        if (error) {
+            throw error;
+        }
+        return result;
+    }
+
+    async $definePluginExports(hostId: string, pluginId: string, proxyNames: string[]): Promise<void> {
+        // tslint:disable-next-line: no-any
+        const pluginManagerInternal = (this.pluginManager as any);
+
+        // add into the activatedPlugins stuff
+        // tslint:disable-next-line: no-any
+        const activatedPlugin: any = {};
+        // tslint:disable-next-line: no-any
+        const proxyExports: any = {};
+        activatedPlugin.exports = proxyExports;
+        const remoteBrowser = this.pluginRemoteBrowser;
+        const deasyncPromise = this.deasyncPromise;
```

Comment:
Not sure what this is for.

Reply:
`this` will be a different object in inner objects.